### ISSUE TITLE
Inconsistency in Sliding Window Pattern in Pretrained LD Model Datasets

### ISSUE THREAD

**Issue by semihkacmaz** [jx-wang-s-group/CoNFiLD#2]
**Description**:

I am trying to train Conditional Neural Field (CNF) and Latent Diffusion (LD) models from scratch for Case 2. While making a step-by-step comparison between my results/model structures and those of the pretrained models you provided, I've discovered an important inconsistency in the LD model dataset structures.

I performed a detailed analysis of the provided pretrained datasets (`train_data.npy` and `valid_data.npy`) to verify if they were created using the sliding window technique. Here's my analysis code:

```python
# Analysis of sliding window pattern in the provided datasets
row_identity = np.all(uncond_train_data[1:, :255, :] == uncond_train_data[:-1, 1:, :], axis=(1, 2))

# Find indices where the pattern breaks
false_indices = np.where(row_identity == False)[0]

# Calculate statistics
percentage_following_pattern = (row_identity.sum() / len(row_identity)) * 100
gaps = np.diff(false_indices)

# Examine the difference magnitudes at break points
def compare_mismatch(data, index):
    diff = data[index+1, :255, :] - data[index, 1:, :]
    return np.sum(np.abs(diff))
```

**Expected Behavior**:
Based on the paper and codebase workflow, it is my understanding that the training and validation datasets should be obtained by applying a sliding window technique to the latents from the CNF model (checkpoint_N.pt). When applying this technique to latents of shape (1200, 256), we should get:
- Training dataset (train_data.npy): (756, 256, 256)
- Validation dataset (valid_data.npy): (189, 256, 256)

**_All consecutive windows should follow the sliding pattern, meaning each row should be equal to the previous row shifted by one position._**

**Actual Behavior**:
Detailed analysis of the pretrained model's dataset reveals unexpected patterns:

1. **Pattern Adherence**:
   - Only 81.19% of rows follow the sliding window pattern (expected percentage 100%)
   - 108 clear breaks in the pattern were identified

2. **Break Characteristics**:
   - Minimum gap between breaks: 1 row
   - Maximum gap between breaks: 27 rows
   - Average gap: 5.33 rows

3. **Difference Magnitude Clusters**:
   - Regular breaks: 600-670 range
   - Medium breaks: 1000-1050 range
   - Large breaks: 1200-1400 range (e.g., 1344, 1353, 1423)

4. **Systematic Patterns**:
   - The difference magnitudes show local consistency
   - Gradual shifts in baseline magnitudes throughout the dataset
   - Periodic occurrence of larger breaks

**Questions**:
1. Is this deviation from the pure sliding window pattern intentional?
2. If yes, what is the reasoning behind these specific patterns in the breaks?
3. If no, could this be affecting the model's performance?
4. Should we modify our data preparation pipeline to match these patterns? If so, how should we modify it? 

Here is the code we are currently using to prepare latents for LD training and validation:

```python
latents = np.load('latents.npy')  # latents extracted from checkpoint_N.pt
window_size = 256
dataset = np.lib.stride_tricks.sliding_window_view(latents, 
                                                 window_shape=(window_size, latents.shape[1]))
train_size = int(0.8 * n_sequences)  # 756
train_dataset = dataset[:train_size]
valid_dataset = dataset[train_size:]
```

**Additional Context**:
The analysis code and detailed results are available if needed. This understanding is crucial for:
- Correctly implementing the data preparation pipeline
- Ensuring model reproducibility
- Understanding potential impacts on model performance

**Environment**:
- Using the latest version of the codebase
- Following the training procedure as described in the documentation
- All dependencies are correctly installed

Would greatly appreciate clarification on this behavior as it's crucial for reproducing the model's performance.

**Labels**: bug, documentation, question

**Comment by meetcfd** [comment]
Hello, 

Thank you for being interested in our work!

So to address your question, your current implementation for the preparing the latents is  **one way**  you can generate the training data for the LD.

In this study, we have chosen the following way for generating the training and valid data: 

```
np.random.seed(42)

latents = np.load('latents.npy')  # latents extracted from checkpoint_N.pt
window_size = 256
dataset = np.stack( [latents[i: i+window_size] for i in range(latents.shape[0]- window_size + 1)] )

data_size = dataset.shape[0]
train_indices = np.random.choice(data_size, int(0.8*data_size), replace=False)
train_test_mask = np.zeros(data_size, dtype=bool)
train_test_mask[train_indices] = True

train_data  = dataset[train_test_mask]
valid_data  = dataset[~train_test_mask]
```

So, the idea behind this snippet of code is to randomly split the latent trajectories into training and validation set.

I hope this explains the source of the distortion observed in provided datasets and resolves your doubt!

**Comment by semihkacmaz** [comment]
Many thanks for the quick and detailed response! That clears up my questions. Appreciate your time. Closing the issue now.
